README.md
# AngularJS Seed

> Olive seed for AngularJS - offers you quickly scaffold an AngularJS project with pragmatic defaults and best practices.

![tech-stack][tech-stack]

## Usage

#### Prepare

Install `gulp`, `bower` and `olive` if you haven't already:
```
npm install -g olive gulp bower
```
Create your project from this Angular seed using Olive CLI:
```
mkdir myapp && cd myapp
olive new angular
```
Install npm and bower dependencies:
```
npm install && bower install
```

#### Develop

Start developing:
```
npm start
```
Run test once and generate code coverage reports in `coverage` directory.
```
npm test
```
Build for production:
```
npm run build
```

## Features

#### Test Driven Development

While developing your app, keep an eye on all unit-test specs.

![reporter][reporter-image]

#### Test coverage

See how much of your code is covered by well-written test scripts.

![coverage][coverage-image]

## Directory Structure

```
├─src/
│ ├─app/
│ │ ├─components/
│ │ │ └─navbar/
│ │ │   ├─repoinfo.directive.js
│ │ │   ├─repoinfo.html
│ │ │   └─repoinfo.scss
│ │ ├─config/
│ │ │ ├─modules.js
│ │ │ ├─routes.js
│ │ │ └─run.js
│ │ ├─layouts/
│ │ │ └─default/
│ │ │   ├─default.html
│ │ │   └─default.scss
│ │ ├─partials/
│ │ │ ├─footer/
│ │ │ │ ├─footer.html
│ │ │ │ └─footer.scss
│ │ │ └─header/
│ │ │   ├─header.html
│ │ │   └─header.scss
│ │ ├─pods/
│ │ │ └─home/
│ │ │   ├─home.controller.js
│ │ │   ├─home.controller.spec.js
│ │ │   ├─home.html
│ │ │   └─home.scss
│ │ ├─services/
│ │ │ └─github
│ │ │   └─github.service.js
│ │ ├─styles/
│ │ │ ├─_overrides.scss
│ │ │ ├─_type.scss
│ │ │ └─app.scss
│ │ └─app.js
│ ├─assets/
│ │ ├─img/
│ │ ├─fonts/
│ │ ├─apple-touch-icon.png
│ │ └─favicon.ico
│ └─index.html
├─dist/
├─.tmp/
├─bower_components/
├─node_modules/
├─.bowerrc
├─.editorconfig
├─.gitattributes
├─.gitignore
├─.jscsrc
├─.jshintrc
├─.oliverc
├─bower.json
├─gulpfile.js
├─karma.conf.js
├─package.json
└─README.md
```

File/Directory    | Purpose
------------------|---------
src/              | Contains your Angular application code.
dist/             | Contains the distributable (that is, optimized and self-contained) output of your application. Deploy this to your server!
.tmp/             | Various temporary output of build steps, as well as the debug output of your application.
bower_components/ |	Bower dependencies.
node_modules      | Node modules required for development purpose.
.bowerrc          | Bower configuration.
.editorconfig     | EditorConfig file.
.oliverc          | Olive configuration.
.gitattributes    | Definition for Git attributes.
.gitignore        | Git configuration for ignored files.
.jscsrc           | JavaScript code style configuration.
.jshintrc         | JSHint configuration.
.oliverc          | Olive configuration.
bower.json        | Bower configuration and dependency list.
gulpfile.js       | Contains build specification for Gulp.
karma.conf.js     | Karma configuration.
package.json      | NPM configuration. Mainly used to list the dependencies needed for asset compilation.
README.md         | This documentation.

## .oliverc

Default options:
```
{
  "paths": {
    "src": "src",
    "tmp": ".tmp",
    "dist": "dist"
  }
}
```
These default options can be overridden. You can also include additional options from the following:

#### Ports

For example:
```
{
  "ports": {
    "app": 3000,
    "bs": 3001,
    "karma": 3002
  }
}
```
- `app`: The app will be served via this port (i.e. http://localhost:3000)
- `bs`: BrowserSync UI control panel can be accessed via this port (i.e. http://localhost:3001)
- `karma`: Karma tests report page is served via this port (i.e. http://localhost:3002/debug.html)

#### Content Security Policy

For example:
```
{
  "content-security-policy": {
    "development": {
      "default-src": "'unsafe-inline' 'self' data: gap: https://ssl.gstatic.com 'unsafe-eval'",
      "style-src": "'self' 'unsafe-inline'",
      "media-src": "*",
      ...
      "other-directive": "other-source"
    },
    "production": {
      "default-src": "'self' data: gap: https://ssl.gstatic.com 'unsafe-eval'",
      "style-src": "'self' 'unsafe-inline'",
      "media-src": "*"
    },
    "other-env": {
      ...
    }
  }  
}
```

[tech-stack]: https://cloud.githubusercontent.com/assets/508043/9190800/b12eb888-3fc7-11e5-97a2-2b1f44a89315.png
[reporter-image]: https://cloud.githubusercontent.com/assets/508043/9190160/172fefc4-3fc0-11e5-97a0-6d52fbb495c3.png
[coverage-image]: https://cloud.githubusercontent.com/assets/508043/9190201/8013b372-3fc0-11e5-94c5-f8cdfcc78b2b.png
[layouts-image]:https://cloud.githubusercontent.com/assets/508043/9190288/77dd3222-3fc1-11e5-891e-be8e1cbf26ed.png


CHANGELOG.md
# Changelog

### 0.2.0

- Upgrade to Angular `1.4.6`.
- Use `ngMaterial` instead of Bootstrap.
- Introduce `jQuery` and `ngTouch`.
- Upgrade Olive to `0.0.8` (use two separate tmp dir for serve and build/test).
- Update `gulpfile` to comply with Olive 0.0.8.
- Update `gulp-useref` search path in `index.html`.
- Remove `layouts` and `partials` and use directives instead for the sake of simplicity of this boilerplate.
- Include Roboto font locally (see in `assets/fonts` dir)


gulpfile.js
/* jshint node:true, latedef:false */

'use strict';

var pkg = require('./package.json'),
    fs = require('fs'),
    olive = require('olive'),
    path = require('path'),
    _ = require('lodash'),
    gulp = require('gulp'),
    gutil = require('gulp-util'),
    chokidar = require('chokidar'),
    wiredep = require('wiredep').stream,
    browserSync = require('browser-sync'),
    modRewrite = require('connect-modrewrite'),
    del = require('del'),
    mkdirp = require('mkdirp'),
    karma = require('karma'),
    tinylr = require('tiny-lr')(),
    opn = require('opn'),
    options,
    $ = require('gulp-load-plugins')({
      pattern: ['gulp-*', 'main-bower-files']
    });

/**
 * Get node environment
 * @return {String} Environment name
 */
function getenv() {
  return process.env.NODE_ENV;
}

/**
 * Set node environment
 * @param  {String} env Environment name
 */
function setenv(env) {
  process.env.NODE_ENV = env;
  getOptions();
}

/**
 * Retrieve olive options
 */
function getOptions() {
  options = olive.getOptions();
}

/**
 * Inject Content-Security-Policy meta tag
 */
function injectCSP() {
  var env = getenv();
  var injectCSPOptions = {
    starttag: '<!-- inject:csp -->',
    transform: function() {
      if (!options.hasOwnProperty('content-security-policy')) { return; }
      var csp = options['content-security-policy'];
      if (!csp.hasOwnProperty(env)) { return; }
      var meta = '<meta http-equiv="Content-Security-Policy" content="';
      var contents = [];
      Object.keys(csp[env]).forEach(function(directive) {
        var source = csp[env][directive];
        contents.push(directive + ' ' + source);
      });
      meta += contents.join('; ') + '">';
      return meta;
    }
  };
  return $.inject(gulp.src(''), injectCSPOptions);
}

/**
 * Inject appinfo script
 */
function injectAppInfo() {
  var injectAppInfoScriptTag = {
    starttag: '<!-- inject:appinfo -->',
    transform: function() {
      var script = '<script src="js/appinfo.js"></script>';
      return script;
    }
  };
  return $.inject(gulp.src(''), injectAppInfoScriptTag);
}

/**
 * Run unit tests
 * @param  {Boolean}  singleRun If true, runs only once
 * @param  {Function} done      Callback function
 */
function runUnitTests(singleRun, done) {
  karma.server.start({
    configFile: path.join(__dirname, 'karma.conf.js'),
    singleRun: singleRun,
    autoWatch: !singleRun,
    reporters: singleRun ? ['mocha', 'coverage'] : ['html']
  }, done);
}

/**
 * Debounced method to reload the unit test report page
 * @param  {String} path Relative path to the changed file
 */
var reloadTestReport = _.debounce(_reloadTestReport, 50);
function _reloadTestReport(path) {
  tinylr.changed({
    body: {
      files: [path]
    }
  });
}

/**
 * Debounced method to reload the browser
 * @param  {String} path Relative path to the changed file
 */
var reloadBrowser = _.debounce(_reloadBrowser, 50);
function _reloadBrowser(path, event) {
  if (event) {
    gutil.log(gutil.colors.cyan('File changed:'), gutil.colors.magenta(path));
  }
  gutil.log(gutil.colors.cyan('Reloading Browsers...'));
  browserSync.reload(path);
}

/**
 * Create `appinfo.js` file in `js` directory under tmp path
 */
gulp.task('appinfo', function() {
  var tmpJsDir = path.join(options.paths.tmp, 'js');
  var appInfoFile = path.join(tmpJsDir, 'appinfo.js');

  var appName = pkg.name,
      appVersion = pkg.version;

  var scriptContent = 'appInfo={name:"' + appName + '",version:"' + appVersion + '"};';
  mkdirp.sync(tmpJsDir);
  fs.writeFileSync(appInfoFile, scriptContent);
});

/**
 * Inject @imports and compile Sass
 */
gulp.task('styles', function() {
  var sassOptions = {
    outputStyle: 'expanded'
  };

  var sassFiles = gulp.src([
    path.join(options.paths.src, 'app/**/*.scss'),
    path.join('!' + options.paths.src, 'app/styles/**/*.scss')
  ], { read: false });

  var injectOptions = {
    transform: function(filePath) {
      filePath = filePath.replace(options.paths.src + '/app/', '../');
      return '@import "' + filePath + '";';
    },
    addRootSlash: false,
    starttag: '// inject:scss',
    endtag: '// endinject'
  };

  return gulp.src(
    path.join(options.paths.src, 'app/styles/app.scss'))
    .pipe($.inject(sassFiles, injectOptions))
    .pipe(wiredep())
    .pipe($.sourcemaps.init())
    .pipe($.sass(sassOptions)).on('error', errorHandler('Sass'))
    .pipe($.autoprefixer()).on('error', errorHandler('Autoprefixer'))
    .pipe($.sourcemaps.write())
    .pipe(gulp.dest(path.join(options.paths.tmp, 'css')));
});

/**
 * Lint the scripts and report issues if any
 */
gulp.task('lint', function() {
  return gulp.src(path.join(options.paths.src, 'app/**/*.js'))
    .pipe($.jshint())
    .pipe($.jscs()).on('error', _.noop)
    .pipe($.jscsStylish.combineWithHintResults())
    .pipe($.jshint.reporter('jshint-stylish'));
});

/**
 * Inject stylesheets into `bower:css` and `inject:css`
 * and scripts into `inject:js`
 */
gulp.task('inject', ['styles', 'lint'], function() {
  var cssFiles = gulp.src([
    path.join(options.paths.tmp, 'css/**/*.css')
  ], { read: false });

  var jsFiles = gulp.src([
    path.join(options.paths.src, 'app/**/*.js'),
    path.join('!' + options.paths.src, 'app/**/*.spec.js')
  ])
  .pipe($.angularFilesort()).on('error', errorHandler('AngularFilesort'));

  // make the path relative
  var injectOptions = {
    ignorePath: [path.join(options.paths.src, 'app'), options.paths.tmp],
    addRootSlash: false
  };

  return gulp.src(path.join(options.paths.src, '*.html'))
    .pipe(injectCSP())                        // inject:csp
    .pipe(injectAppInfo())                    // inject:appinfo
    .pipe($.inject(cssFiles, injectOptions))  // inject:css
    .pipe($.inject(jsFiles, injectOptions))   // inject:js
    .pipe(wiredep())                          // bower:css
    .pipe(gulp.dest(options.paths.tmp));

  /* Important: While using both, `wiredep` should always run after `inject` in the pipe */
});

/**
 * Watch for changes
 */
gulp.task('watch', ['inject'], function() {

  // Watch for changes in the root htmls (i.e. index.html) and in bower.json
  chokidar.watch([path.join(options.paths.src, '*.html'), 'bower.json'], {ignoreInitial: true}).on('change', function(_path) {
    gulp.start('inject', function() {
      reloadBrowser(_path, 'change');
      reloadTestReport(_path, 'change');
    });
  });

  // Watch for changes in all files inside the assets directory
  chokidar.watch(path.join(options.paths.src, 'assets'), {ignoreInitial: true}).on('all', function(event, _path) {
    reloadBrowser(_path, event);
  });

  // Watch for changes in scss/js/html files inside app directory
  chokidar.watch(path.join(options.paths.src, 'app'), {ignoreInitial: true}).on('all', function(event, _path) {
    var ext = path.extname(_path);

    // Watch for change in `app/**/*.scss`
    if (ext === '.scss') {
      if (event === 'change') {
        gulp.start('styles', function() {
          reloadBrowser(_path, event);
        });
      } else {
        gulp.start('inject', function() {
          reloadBrowser(_path, event);
        });
      }
    }

    // Watch for changes in `app/**/*.js`
    else if (ext === '.js') {
      if (event === 'change') {
        gulp.start('lint', function() {
          reloadBrowser(_path, event);
          reloadTestReport(_path);
        });
      } else {
        gulp.start('inject', function() {
          reloadBrowser(_path, event);
          reloadTestReport(_path);
        });
      }
    }

    // Watch for change in `app/**/*.html`
    else if (ext === '.html') {
      reloadBrowser(_path, event);
      reloadTestReport(_path);
    }

  });

  // Listen to livereload update for html report page
  tinylr.listen(35729);

  // Run unit test continuously
  runUnitTests(false);

  // Open report page in browser
  opn('http://localhost:' + options.ports.karma + '/debug.html');
});

/**
 * Serve via Browsersync
 */
gulp.task('serve', ['setenv:development', 'clean:tmp', 'appinfo', 'watch'], function() {
  var ports = options.ports || {};
  browserSync.init({
    port: ports.app || 3000,
    ui: {
      port: ports.bs || 3001
    },
    server: {
      baseDir: [options.paths.tmp, path.join(options.paths.src, 'assets'), path.join(options.paths.src, 'app')],
      routes: {
        '/bower_components': 'bower_components'
      },
      middleware: [
        modRewrite(['!\\.\\w+$ /index.html [L]']),
        function (req, res, next) {
          res.setHeader('X-UA-Compatible', 'IE=Edge');
          next();
        }
      ]
    },
    notify: false,
    logLevel: 'silent'
  });
});

/**
 * Run unit tests once
 */
gulp.task('test', ['setenv:test', 'lint'], function(done) {
  runUnitTests(true, done);
});

/**
 * Run unit tests continuously
 */
gulp.task('test:auto', ['setenv:test', 'watch'], function() {
  runUnitTests(false);
});

/**
 * Delete tmp directory
 */
gulp.task('clean:tmp', function() {
  del.sync(options.paths.tmp);       // delete tmp dir
  mkdirp.sync(options.paths.tmp);    // recreate it (empty)
});

/**
 * Delete dist directory
 */
gulp.task('clean:dist', function() {
  del.sync([
    options.paths.dist + '/**/*',
    '!' + options.paths.dist + '/.gitkeep'
  ], { dot: true });
});

/**
 * Delete both tmp and dist directories
 */
gulp.task('clean', function() {
  gulp.start('clean:tmp');
  gulp.start('clean:dist');
});

/**
 * Set environment to development
 */
gulp.task('setenv:development', function() {
  setenv('development');
});

/**
 * Set environment to production
 */
gulp.task('setenv:production', function() {
  setenv('production');
});

/**
 * Set environment to test
 */
gulp.task('setenv:test', function() {
  setenv('test');
});

/**
 * Convert all angular html templates into a javascript template cache
 */
gulp.task('templates', function() {
  return gulp.src(path.join(options.paths.src, 'app/**/*.html'))
    .pipe($.minifyHtml({
      empty: true,
      spare: true,
      quotes: true
    }))
    .pipe($.angularTemplatecache({
      module: 'app'
    }))
    .pipe(gulp.dest(path.join(options.paths.tmp, 'js')));
});

/**
 * Build app html, css and js files
 */
gulp.task('app', ['inject', 'templates'], function() {
  var templateFiles = gulp.src(path.join(options.paths.tmp, 'js/templates.js'), { read: false });
  var injectOptions = {
    starttag: '<!-- inject:templates -->',
    ignorePath: options.paths.tmp,
    addRootSlash: false
  };

  var htmlFilter = $.filter('*.html');
  var cssFilter = $.filter('**/*.css');
  var jsFilter = $.filter('**/*.js');
  var assets = $.useref.assets();

  return gulp.src(path.join(options.paths.tmp, '*.html'))
    .pipe(injectCSP())
    .pipe($.inject(templateFiles, injectOptions))
    .pipe(assets)
    .pipe($.rev())
    .pipe(jsFilter)
    .pipe($.ngAnnotate())
    .pipe($.uglify().on('error', errorHandler('Uglify')))
    .pipe(jsFilter.restore())
    .pipe(cssFilter)
    .pipe($.minifyCss({
      processImport: false,
      keepSpecialComments: false
    }))
    .pipe(cssFilter.restore())
    .pipe(assets.restore())
    .pipe($.useref())
    .pipe($.revReplace())
    .pipe(htmlFilter)
    .pipe($.minifyHtml({
      empty: true,
      spare: true,
      quotes: true,
      conditionals: true
    }))
    .pipe(htmlFilter.restore())
    .pipe(gulp.dest(options.paths.dist))
    .pipe($.size({
      title: options.paths.dist,
      showFiles: true
    }));
});

/**
 * Copy fonts from bower dependencies to dist/fonts
 * Note: Custom fonts are handled by `assets` task
 */
gulp.task('fonts', function () {
  return gulp.src($.mainBowerFiles())
    .pipe($.filter('**/*.{eot,svg,ttf,woff,woff2}'))
    .pipe($.flatten())
    .pipe(gulp.dest(path.join(options.paths.dist, 'fonts')));
});

/**
 * Copy content of the assets directory
 */
gulp.task('assets', function() {
  return gulp.src(path.join(options.paths.src, 'assets/**/*'))
    .pipe(gulp.dest(options.paths.dist));
});

/**
 * Build task
 */
gulp.task('build', [
  'setenv:production',
  'clean',
  'appinfo',
  'app',
  'fonts',
  'assets'
]);

/**
 * Default task
 */
gulp.task('default', [
  'build'
]);

/**
 * Common implementation for an error handler of a gulp plugin
 * @param  {String}   title Short description of the error
 * @return {Function}       Callback function
 */
function errorHandler(title) {
  return function(err) {
    gutil.log(gutil.colors.red('[' + title + ']'), err.toString());
    this.emit('end');
  };
}
